### ISSUE TITLE
result visualization and regarding f1 score

### ISSUE THREAD

**Issue by doheon114** [bbj-lab/protoecgnet#5]
Hi, I have a couple more questions, if you don’t mind.

First, based on your paper, it seems that when a test sample is input, the model is able to retrieve the most similar prototype from the training data, as illustrated in the figure below:

<img width="883" alt="Image" src="https://github.com/user-attachments/assets/2ba39177-ef82-4e84-acd3-f98af6d7c0b8" />
However, after reviewing `ecg_utils.py`, I couldn't find any implementation for this prototype retrieval functionality. Could you kindly provide an example code snippet or clarify how this feature can be replicated?

Secondly, I noticed that your model achieved a good AUROC score on the test data, which is impressive. However, after the fusion step, I observed a relatively low F1 score — approximately 0.1XX — in my own experiments. I'm wondering if this is a known limitation of the model, or if it might indicate an issue with how I conducted the experiments. I also noticed that the F1 score wasn’t reported in your paper, so I would really appreciate it if you could share any insights regarding your F1 score results.

**Comment by sahilsethi0105** [comment]
Hello, thank you for reaching out. Regarding your questions: 

1. The code used to produce the case-based explanation figures you reference was not originally intended for public release. However, we will be uploading a Jupyter notebook in the near future with this functionality. I will leave this issue open until we have posted it. That being said, the prototype images (and the rhythm strips for 1D rhythm branch, and the time-localized cut-outs for 2D morphology branch) are all automatically produced and saved by ```push.py```. I will provide a code-snippet below to find the most activated prototype for a given test ECG (but note that, for case-based explanations, we show the top-activated prototype for each predicted class, not necessarily the top-activated prototype).

```
# Load projection metadata (prototype assignments) 
with open(json_path, "r") as f:
    prototype_metadata = json.load(f)

prototype_class_mapping = {
    int(proto_id): meta["prototype_class"]
    for proto_id, meta in prototype_metadata.items()
}

# Inference outputs
probabilities, similarity_scores = run_inference(model, X_sample)

def get_top_prototypes(similarity_scores, top_k=5):
    top_indices = np.argsort(similarity_scores)[-top_k:][::-1]
    top_scores = similarity_scores[top_indices]
    return list(zip(top_indices, top_scores))

print("\n Top Activated Prototypes:")
top_protos = get_top_prototypes(similarity_scores, top_k=top_k)

for i, (proto_idx, score) in enumerate(top_protos):
    class_name = prototype_class_mapping.get(proto_idx, "Unknown")
    print(f"  {i+1}. Prototype {proto_idx} ({class_name}) – Score: {score:.4f}")
```



2. Regarding F1 scores, this is not a metric usually reported when conducting multi-label classification on the full 71-label set of PTB-XL. This is due to several factors, especially including class-imbalance and the clinical need to tune thresholds per-class/diagnosis. However, I believe our average F1 scores across all 71 labels were in the 0.30-0.40 range whether using the ResNet blackbox or ProtoECGNet, so I am not sure why yours would be so low. I would suggest hyperparameter-tuning each of the training stages, as we do in our paper, if you have not already. 

Please let me know if you have any more questions!

**Comment by doheon114** [comment]
1. Thank you for sharing !! I will try it later.
2. Ok. I will recheck about this problem.

**Comment by MahmoudAwad7** [comment]
Hello, and thank you for releasing ProtoECGNet!

I'm currently testing the inference part of the model using your public repo. However, the repository does not include:

- the trained `.pth` weights  
- the `standard_scaler_2D.pkl` file  
- any precomputed label co-occurrence matrices  

Would it be possible for you to share:
1. A pretrained model (e.g., 2D model trained on label set 3 with custom_groups=True)  
2. The corresponding fitted StandardScaler  
3. Any necessary files for inference (if required)  

This would be extremely helpful for reproducibility and testing.  
Thanks again for the great work!

**Comment by sahilsethi0105** [comment]
Hi Mahmoud, 

Thank you for reaching out! 

1. I recommend training from scratch following the instructions in the ReadMe. It seems that others (as in the issue above) have been able to train models using the repo. Our focus was on developing a method that can be applied to different datasets, rather than developing a foundation model for others to use out-of-the-box, so we currently do not plan to release weights. 
2. StandardScaler is not used in our current model, as we do not standardize the input ECGs. If you set standardize=True, then the code will automatically compute and store this .pkl file for you
3. As above, weights will likely not be released, but we plan to release a Jupyter notebook that helps walk you through producing the case-based explanations in the future. However, I have already provided a code snippet in my response above that shows how to determine the most activated prototypes and their similarity scores at inference

If you run into any specific issues with training, please let me know the exact steps you took and error messages you received, and I’d be happy to help. 

Best, 
Sahil Sethi

**Comment by MahmoudAwad7** [comment]
Hi Sahil,

Thank you so much for your detailed response and clarification!

I'll proceed with training from scratch as suggested, and I really appreciate your willingness to help if I run into any issues during the process.

Looking forward to the upcoming Jupyter notebook for case-based explanations — that sounds very helpful!

Best regards,
Mahmoud

**Comment by sahilsethi0105** [comment]
Closing this issue for now as we just added [`src/case_explanations.ipynb`](https://github.com/bbj-lab/protoecgnet/blob/main/src/case_explanations.ipynb) and [`src/inference_fusion.py`](https://github.com/bbj-lab/protoecgnet/blob/main/src/inference_fusion.py) to the repo. The ReadMe has been updated with instructions.
